src/ray/raylet/agent_manager.cc

```diff
@@ -128,15 +128,25 @@ void AgentManager::StartAgent() {
 
     int exit_code = child.Wait();
     timer->cancel();
-    RAY_LOG(WARNING) << "Agent process with id " << agent_id << " exited, return value "
-                     << exit_code << ". ip " << reported_agent_ip_address_ << ". id "
-                     << reported_agent_id_;
-    RAY_LOG(ERROR)
-        << "The raylet exited immediately because the Ray agent failed. "
-           "The raylet fate shares with the agent. This can happen because the "
-           "Ray agent was unexpectedly killed or failed. See "
-           "`dashboard_agent.log` for the root cause.";
-    QuickExit();
+    RAY_LOG(INFO) << "Agent process with id " << agent_id << " exited, exit code "
+                  << exit_code << ". ip " << reported_agent_ip_address_ << ". id "
+                  << reported_agent_id_;
+
+    if (exit_code == 0) {
+      RAY_LOG(INFO) << "Agent is terminated gracefully (exit code 0). "
+                    << "Raylet will shutdown gracefully.";
+      // Sending a SIGTERM to itself is equivalent to gracefully shutting down raylet.
+      RAY_CHECK(std::raise(SIGTERM) == 0) << "There was a failure while sending a "
+                                             "sigterm to itself. The process will not "
+                                             "gracefully shutdown.";
+    } else {
+      RAY_LOG(ERROR)
+          << "The raylet exited immediately because the Ray agent failed. "
+             "The raylet fate shares with the agent. This can happen because the "
+             "Ray agent was unexpectedly killed or failed. See "
+             "`dashboard_agent.log` for the root cause.";
+      QuickExit();
```

Comment:
How about this error code propogation instead: 
- if agent SIGTERM, it exits with SIGTERM
- raylet's QuickExit exits with the agent's exit code if it's non-zero. 

I think this should: 
1. Have both agent and raylets exit with SIGTERM (or other exit codes), which is also consistent with the fate-sharing (sharing the same exit code) 
2. Graceful shutdown of both if it's normal. 

We are handling SIGTERMs anyway at the `ray stop --blocking` call there, I feel we should interfere less with the components on what exitcodes they surface.

Reply:
Hmm what about we just detect exitcode==SIGTERM instead of QuickExit(SIGTERM)? I think doing graceful shutdown makes more sense in this case.

Reply:
oh i see, yeah, SGTM.

Reply:
Btw I am going to add an additional unit test to check the exit code of raylet upon agent shutdown